node/lib/util/status.js

```diff
@@ -43,235 +53,411 @@ const SubmoduleUtil       = require("../util/submodule_util");
 const SubmoduleConfigUtil = require("../util/submodule_config_util");
 
 /**
- * Return a string describing the file changes in the specified `repoStatus` or
- * an empty string if there are no changes.
- *
- * @param {RepoStatus} repoStatus
- * @return {String}
+ * This value-semantic class describes a line entry to be printed in a status
+ * message.
  */
-exports.printFileStatuses = function (repoStatus) {
-    assert.instanceOf(repoStatus, RepoStatus);
-    let result = "";
-    const FILESTATUS = RepoStatus.FILESTATUS;
-    function statusDescription(status) {
-        switch(status) {
+class StatusDescriptor {
+    /**
+     * @param {RepoStatus.FILESTATUS} status
+     * @param {String}                path
+     * @param {String}                detail
+     */
+    constructor(status, path, detail) {
+        this.status = status;
+        this.path = path;
+        this.detail = detail;
```

Comment:
I didn't put a check in for these because I've found that the assertions can be prohibitively expensive, and I expect that there may be a great number of these created. For example, the `assert.isObject` at the beginning of the constructor for `RepoAST.Submodule` cost 500ms when invoked 26k times.

I wish JS (and Java) had macros.

I guess I could put a comment explaining why they aren't there, at least.